8285868: x86_64 intrinsics for floating point methods isNaN, isFinite, IsInfinite

vamsi-parasa opened the pull request:
We develop optimized x86_64 intrinsics for the floating point class check methods isNaN(), isFinite() and IsInfinite() for Float and Double classes. JMH benchmarks show ~8x improvement for isNan(), ~3x improvement for isInfinite() and 15% gain for isFinite().

vamsi-parasa commented:
/label hotspot-compiler

vnkozlov reviewed (commented):
Impressive. Few comments.

You are testing performance of storing `boolean` results into array but usually these Java methods used in conditions. Measuring that will be more real word case. For both case: with `avx512dq` On and OFF.

And you need to post you perf results at least in RFE. Please, also show what instructions are currently generated vs your changes. I don't get how you made `isNaN()` faster - you generate more instructions is seems.

Instead of 3 new Ideal nodes per type you can use one and store instrinsic id (or other enum) in its field which you can read in `.ad` file instructions. Instead I suggest to split those mach instructions based on `avx512dq` support to avoid unused registers killing.

Why Double type support is limited to LP64? Why there is no `x86_32.ad` changes?

You can reuse `tmp1` in `double_class_check()`.

merykitty commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -4740,3 +4740,101 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
+  cmpl(temp, POS_INF);
+  switch (opcode) {
+    case Op_IsFiniteF:
+      setb(Assembler::below, dst);
```
This partial write may stall later reads on `dst`, you could emit a `xor dst, dst` before doing the comparison.

vamsi-parasa replied:
Will try the `xor dst, dst` and see the performance changes.

vamsi-parasa replied:
setb either writes a 0 or 1 based on the condition. So, will it cause partial writes?

merykitty replied:
Actually `setb` only writes the byte portion and leaves the remaining of the register intact, so it would be wrong without clearing the register beforehand.

vamsi-parasa replied:
`setb` is producing the correct results and also adding the `xor dst, dst` didn't give any performance improvement. Is it still necessary?

jatin-bhateja replied:
> This partial write may stall later reads on `dst`, you could emit a `xor dst, dst` before doing the comparison.

APIs generate boolean result,  and any reader should only be consuming 8 bit result. Thus its not a scenario for partial register stall from HW perspective,  a narrow write followed by wider read cause a partial register stall due to additional cycles penalty incurred while re-assembling the unmodified bits with modified portion.

merykitty replied:
@vamsi-parasa Other boolean producers such as `Conv2B` or `VectorTest` widen the byte value to 32 bit after `setb` so I believe a zeroing is necessary here. You accidentally achieve correct results in tests because the boolean stores only read the least significant 8 bits of the `dst` registers. Other operations such as `test` will read full `int` and may lead to incorrect results.

@jatin-bhateja Boolean is just dressed-up int except in stores so most readers will consume 32 bits of the register.

Thanks.

vamsi-parasa replied:
@merykitty You might be right. `setb` was causing a wierd jdk build error in `java.lang.HashMap` due to to error in `isNaN()`. It got fixed by clearing the higher bytes of `dst` register which is being set by `setb`.

jatin-bhateja replied:
> @vamsi-parasa Other boolean producers such as `Conv2B` or `VectorTest` widen the byte value to 32 bit after `setb` so I believe a zeroing is necessary here. You accidentally achieve correct results in tests because the boolean stores only read the least significant 8 bits of the `dst` registers. Other operations such as `test` will read full `int` and may lead to incorrect results.
> 
> @jatin-bhateja Boolean is just dressed-up int except in stores so most readers will consume 32 bits of the register.
> 
> Thanks.

For explicit widening from byte to int compiler will emit an explicit zero extension instruction (movzbl).  There will not be any case where byte will be directly read along with unmodified bits  a consumer.

merykitty commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -4740,3 +4740,101 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
+  movq(temp, src);
+  mov64(temp1, KILL_SIGN_MASK);
+  andq(temp, temp1);
+  mov64(temp2, POS_INF);
```
Can we use `temp1` for this, too?

vamsi-parasa replied:
Sure, will make the change

merykitty commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -4740,3 +4740,101 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
+  vfpclassss(tmp, src, imm8);
+  kmovbl(dst, tmp);
+  if (opcode == Op_IsFiniteF) {
+    xorl(dst, 0x00000001);
```
`notl(dst)`?

vamsi-parasa replied:
I can't recall why `notl(dst)` didn't work. Will try it and let you know.

vamsi-parasa replied:
notl(dst) doesn't work. we need to flip only the last bit as the kreg stores the result in the last bit.

merykitty reviewed (commented):
Hi, I'm working on #8525 which also improves the performance of these methods. With that patch, `isNaN` is reduced to the optimal sequence `ucomiss x, x; jp label`. This patch still benefits the performance of `isFinite` and `isInfinite` for float cases. For double cases I'm not sure due to the materialisation of long constants, though.

Also, can we output the result of the intrinsics directly in the flag registers? Thanks.

merykitty commented:
Also, for non `vfpclass` cases, it would be simpler and more efficient to implement in the Java side instead

    static Float::isFinite(float f) {
        return (floatToRawIntBits(f) & SIGN_ELIMINATION) < POS_INFINITY_BITS;
    }

vamsi-parasa commented:
> Hi, I'm working on #8525 which also improves the performance of these methods. With that patch, `isNaN` is reduced to the optimal sequence `ucomiss x, x; jp label`. This patch still benefits the performance of `isFinite` and `isInfinite` for float cases. For double cases without `vfpclass`, I'm not sure due to the materialisation of long constants, though.
> 
> Also, can we output the result of the intrinsics directly in the flag registers? Thanks.

Thank you for the review! The main bottleneck for the performance of `isNan()` comes from popfq instruction used for fixing up the bits in flags register. Is your patch #8525 fixing that?

vamsi-parasa commented:
> Also, for non `vfpclass` cases, it would be simpler and more efficient to implement in the Java side instead
> 
> ```
> static Float::isFinite(float f) {
>     return (floatToRawIntBits(f) & SIGN_ELIMINATION) < POS_INFINITY_BITS;
> }
> ```

True, but that needs changes to java.lang.Float. Will that be approved?

vamsi-parasa commented:
> Impressive. Few comments.
> 
> You are testing performance of storing `boolean` results into array but usually these Java methods used in conditions. Measuring that will be more real word case. For both case: with `avx512dq` On and OFF.
> 
> And you need to post you perf results at least in RFE. Please, also show what instructions are currently generated vs your changes. I don't get how you made `isNaN()` faster - you generate more instructions is seems.
> 
> Instead of 3 new Ideal nodes per type you can use one and store instrinsic id (or other enum) in its field which you can read in `.ad` file instructions. Instead I suggest to split those mach instructions based on `avx512dq` support to avoid unused registers killing.
> 
> Why Double type support is limited to LP64? Why there is no `x86_32.ad` changes?
> 
> You can reuse `tmp1` in `double_class_check()`.

Hi Vladimir (@vnkozlov),
Sorry for the delay! 
As per your suggestions, the JMH benchmarks were updated to use these Java methods in conditions and updated the RFE with performance data with and without the vfpclassss/d instructions. Also removed the redundant temp2 as per your suggestion.
Will work on condensing the 3 nodes to one and add support for x86_32.ad.

Thanks,
Vamsi

jatin-bhateja commented on src/hotspot/share/opto/fpclassnode.hpp:
```diff
@@ -0,0 +1,86 @@
+  virtual uint ideal_reg() const { return Op_RegI; }
+};
+
+//---------- IsNaNFNode -----------------------------------------------------
+class IsNaNFNode : public Node {
+  public:
+  IsNaNFNode(Node* in1) : Node(0, in1) {}
+  virtual int   Opcode() const;
+  const Type* bottom_type() const { return TypeInt::BOOL; }
+  virtual uint ideal_reg() const { return Op_RegI; }
+};
+
+//---------- IsInfiniteDNode -----------------------------------------------------
+class IsInfiniteDNode : public Node {
+  public:
+  IsInfiniteDNode(Node* in1) : Node(0, in1) {}
+  virtual int   Opcode() const;
+  const Type* bottom_type() const { return TypeInt::BOOL; }
+  virtual uint ideal_reg() const { return Op_RegI; }
+};
+
+//---------- IsFiniteDNode -----------------------------------------------------
+class IsFiniteDNode : public Node {
+  public:
+  IsFiniteDNode(Node* in1) : Node(0, in1) {}
+  virtual int   Opcode() const;
+  const Type* bottom_type() const { return TypeInt::BOOL; }
+  virtual uint ideal_reg() const { return Op_RegI; }
+};
+
+//---------- IsNaNDNode -----------------------------------------------------
+class IsNaNDNode : public Node {
+  public:
+  IsNaNDNode(Node* in1) : Node(0, in1) {}
+  virtual int   Opcode() const;
+  const Type* bottom_type() const { return TypeInt::BOOL; }
+  virtual uint ideal_reg() const { return Op_RegI; }
+};
+
+#endif // SHARE_OPTO_FPCLASSNODE_HPP
```
You can move these to src/hotspot/share/opto/intrinsicnode.hpp

vamsi-parasa replied:
Thanks Jatin! Will make the change and push the update.

merykitty commented:
@vamsi-parasa  Yes #8525 change the matching of `Bool` to `cmpOpUCF` for eq/ne where both inputs are the same and add `CMove` rules for `cmpOpUCF2`, which should prevent floating point comparison from matching `cmpOpU`, which has bad overhead of fixing the flags.

jatin-bhateja commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -11120,6 +11120,67 @@ instruct MoveL2D_reg_reg(regD dst, rRegL src) %{
+  ins_pipe(pipe_slow);
+%}
+
+instruct DoubleClassCheck_reg_reg_vfpclass(rRegI dst, regD src, kReg ktmp, rFlagsReg cr)
+%{
+  predicate(VM_Version::supports_avx512dq());
+  match(Set dst (IsInfiniteD src));
+  match(Set dst (IsNaND src));
+  match(Set dst (IsFiniteD src));
+  effect(TEMP ktmp, KILL cr);
+  format %{ "double_class_check $dst, $src" %}
+  ins_encode %{
+    int opcode = this->ideal_Opcode();
+      __ double_class_check_vfp(opcode, $dst$$Register, $src$$XMMRegister, $ktmp$$KRegister);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct DoubleClassCheck_reg_reg(rRegI dst, regD src, rRegL tmp, rRegL tmp1, rFlagsReg cr)
+%{
+  predicate(!VM_Version::supports_avx512dq());
+  match(Set dst (IsInfiniteD src));
+  match(Set dst (IsNaND src));
+  match(Set dst (IsFiniteD src));
+  effect(TEMP tmp, TEMP tmp1, KILL cr);
+  format %{ "double_class_check $dst, $src" %}
+  ins_encode %{
+    int opcode = this->ideal_Opcode();
+    __ double_class_check(opcode, $src$$XMMRegister, $dst$$Register, $tmp$$Register,
+                $tmp1$$Register);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
```
Should this be guarded by LP64 ? since associated macro assembly routines have that check ?

Also match_rule_supported vector should prevent intrinsifying these for 32 bit platform if you do not plan to handle them.

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/DoubleClassCheck.java:
```diff
@@ -0,0 +1,89 @@
+    @OperationsPerInvocation(BUFFER_SIZE)
+    public void testIsInfinite() {
+        for (int i = 0; i < BUFFER_SIZE; i++) {
+            outputs[i] = Double.isInfinite(inputs[i]) ? false : true;
```
Any specific reason to explicitly add conditional check to return true/false when isInfinit returns bool value ?

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/DoubleClassCheck.java:
```diff
@@ -0,0 +1,89 @@
+    @OperationsPerInvocation(BUFFER_SIZE)
+    public void testIsNaN() {
+        for (int i = 0; i < BUFFER_SIZE; i++) {
+            outputs[i] = Double.isNaN(inputs[i]) ? false : true;
```
Any specific reason to explicitly add conditional check to return true/false when isNaN returns bool value ?

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/DoubleClassCheck.java:
```diff
@@ -0,0 +1,89 @@
+    @OperationsPerInvocation(BUFFER_SIZE)
+    public void testIsFinite() {
+        for (int i = 0; i < BUFFER_SIZE; i++) {
+            outputs[i] = Double.isFinite(inputs[i]) ? false : true;
```
Any specific reason to explicitly add conditional check to return true/false when isFinite returns bool value ?

vamsi-parasa replied:
Initially, it was returning a boolean value and storing it in the output buffer. Vladimir suggested that the realword usecase of these methods is in conditions. Hence, the benchmarks were modified.

jatin-bhateja replied:
Yes, that makes sense, but being a micro benchmark we micro focusing on perf gain due to this particular API, may be you can have one stand alone case also.

vamsi-parasa replied:
Sure, will add the standalone case also.

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/FloatClassCheck.java:
```diff
@@ -0,0 +1,90 @@
+    @OperationsPerInvocation(BUFFER_SIZE)
+    public void testIsFinite() {
+        for (int i = 0; i < BUFFER_SIZE; i++) {
+            outputs[i] = Float.isFinite(inputs[i]) ? false : true;
```
Same a above

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/FloatClassCheck.java:
```diff
@@ -0,0 +1,90 @@
+    @OperationsPerInvocation(BUFFER_SIZE)
+    public void testIsInfinite() {
+        for (int i = 0; i < BUFFER_SIZE; i++) {
+            outputs[i] = Float.isInfinite(inputs[i]) ? false : true;
```
Same a above

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/FloatClassCheck.java:
```diff
@@ -0,0 +1,90 @@
+    @OperationsPerInvocation(BUFFER_SIZE)
+    public void testIsNaN() {
+        for (int i = 0; i < BUFFER_SIZE; i++) {
+            outputs[i] = Float.isNaN(inputs[i]) ? false : true;
```
Same as above

jatin-bhateja commented on test/micro/org/openjdk/bench/java/lang/FloatClassCheck.java:
```diff
@@ -0,0 +1,90 @@
+ */
+package org.openjdk.bench.java.lang;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Param;
+import java.util.random.RandomGenerator;
+import java.util.random.RandomGeneratorFactory;
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Thread)
+@Fork(1)
+public class FloatClassCheck {
+
+    RandomGenerator rng;
```
Just a suggestion we can also create one benchmark to handle both the floating point types.

vamsi-parasa replied:
True, I thought keeping them separate would allow future tests could be added for each type.

vnkozlov replied:
Yes, please keep them separate.

jatin-bhateja commented on src/hotspot/share/opto/fpclassnode.hpp:
```diff
@@ -0,0 +1,86 @@
+  public:
+  IsInfiniteDNode(Node* in1) : Node(0, in1) {}
+  virtual int   Opcode() const;
+  const Type* bottom_type() const { return TypeInt::BOOL; }
+  virtual uint ideal_reg() const { return Op_RegI; }
```
None of the IR nodes handle constant folding scenarios.

vamsi-parasa commented:
> @vamsi-parasa Yes #8525 change the matching of `Bool` to `cmpOpUCF` for eq/ne where both inputs are the same and add `CMove` rules for `cmpOpUCF2`, which should prevent floating point comparison from matching `cmpOpU`, which has bad overhead of fixing the flags.

Glad to know that you removed the overhead of fixing up the flags register.

vamsi-parasa commented:
> Please remove `isNaN` intrinsics in favour of #8525 .
> 
> Also, you should not use `andl(dst, 0xff)` to zero out the upper bits of `dst` since it is a 32-bit read following a 8-bit write, constitute a partial register stall
> 
> Refer to section 3.5.2.4, Partial register stalls from Intel® 64 and IA-32 Architectures Optimization Reference Manual:
> 
> > A partial register stall happens when an instruction refers to a register, portions of which were previously modified by other instructions.
> 
> There are 2 options worth considering here
> 
> * Zeroing the register before `setb` instruction, referring to the same section
>   > For optimal performance, use of zero idioms, before the use of the register, eliminates the need for partial register
>   > merge micro-ops
>   
>   
>   This is more preferable since it does not contribute an execution uop in the backend (but still consumes a slot in the
>   decoder and uop cache)
> * Zero extending the register after the `setb` instruction. This is less optimal since it has an extra latency of zero extension and adding a real executed uop in the backend.
> 
> Thanks.

#8525 seems to be eliminating the flags register fixup for `IsNaN()`. These intrinsics can show speedup over `ucomiss` instructions. Also, having the intrinsic can be used for future vectorization. So, we can keep the `IsNaN()` intrinsic along with your improvement. Both are orthogonal, not mutually exclusive. 

Actually, `andl(dst, 0xff)` is giving speedup over zeroing out the register before `setb`. Also, would a 32bit logical `and` of all bits not the problem you mentioned?

merykitty commented:
It sounds strange, please show the asm of your patch with respects to the benchmark. Also, please try cmoving with other arbitrary values such as 19 and 7 instead of `false` and `true`. The latter may be recognised as simple boolean not operation, remove the real comparison part, which defeats the purpose of Vladimir's suggestion.

Regarding vectorisation, `isNaN` is a simple comparison and can be easily auto-vectorised without help from intrinsics.

My speculation:

A native comparison such as `x != x` can be parsed directly by the compiler. As a result, the graph of the expression `if (x != x)` is simply

    CmpF
     |
    Bool
     |
     If

Your intrinsics, on the other hand, do not return the results on the flags, which leads to an extra comparison when using in conditions, `if(isNaN(x))` becomes

    IsNaN    0
        \   /
        CmpI
         |
        Bool
         |
         If

In your benchmark, however, using this comparison to cmoving between 0 and 1 (`false` and `true`), the compiler recognised the pattern `x != 0 ? 0 : 1` with `x` having the type of `TypeInt::BOOL`. As a result, it reduces the graph into

    IsNaN    1
        \   /
        XorI

Personally, I'm not into this implementation of intrinsics. FYI, gcc and clang both use sequences similar to `x != x` for `std::isnan`, `Math.abs(x) <= MAX_VALUE` for `std::isfinite` and `Math.abs(x) > MAX_VALUE` for `std::isinf`. The first one reduces to a single instruction `ucomiss x, x` so there is no reason to optimise further. The others are compiled down to 2 instructions each `vandpd t, x, [SIGN_ELIMINATE]; ucomiss t, [MAX_VALUE]`, so to optimise these further requires careful assessments.

If you feel comfortable I would suggest you build the graph for these intrinsics as

     X
     |
    Bool   1    0
       \   |   /
         CMove

Then we can add ideal rules to `BoolNode` to recognise the patterns

     X
     |
    Bool   1    0
       \   |   /
         CMove     0
              \   /
               CmpI
                |
               Bool

And reduce them to

     X
     |
    Bool

With this, we can have the `Double::isInfinite` intrinsics compiled down to `vfpclass k, x; ktest k`, which is much more preferable. For non-AVX512DQ though I would prefer implementing them in Java similar to described above. Both abs and comparison nodes are not hard to be vectorised so it would not be a problem.

Thanks a lot.

jatin-bhateja commented:
> * Zeroing the register before `setb` instruction, referring to the same section
>   > For optimal performance, use of zero idioms, before the use of the register, eliminates the need for partial register
>   > merge micro-ops
>   
>   
>   This is more preferable since it does not contribute an execution uop in the backend (but still consumes a slot in the
>   decoder and uop cache)

Agree.  even a prior, movl REG, 0 can be used, it never reaches execution ports.

jatin-bhateja commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -4740,3 +4740,101 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
+  int32_t POS_INF = 0x7f800000;
+  int32_t KILL_SIGN_MASK = 0x7fffffff;
+
+  movdl(temp, src);
+  andl(temp, KILL_SIGN_MASK);
+  cmpl(temp, POS_INF);
```
For IsNaN following sequence will offer better latency
     "vucomiss src_xmm, src_xmm"
     "setp r8"

jatin-bhateja replied:
Hi @vamsi-parasa ,

I can see almost 30% improvement in you JMH micro for IsNaN through above suggested change.
Original score over AVX2 machine:

Benchmark                       Mode  Cnt  Score   Error  Units
FloatClassCheck.testIsNaN       avgt    2  0.854          ns/op
FloatClassCheck.testIsNaN:·asm  avgt         NaN            ---

With new sequence: 
Benchmark                       Mode  Cnt  Score   Error  Units
FloatClassCheck.testIsNaN       avgt    2  0.570          ns/op
FloatClassCheck.testIsNaN:·asm  avgt         NaN            ---

vamsi-parasa replied:
Thanks Jatin! Will test it on my machine as well...

vamsi-parasa commented:
> It sounds strange, please show the asm of your patch with respects to the benchmark. Also, please try cmoving with other arbitrary values such as 19 and 7 instead of `false` and `true`. The latter may be recognised as simple boolean not operation, remove the real comparison part, which defeats the purpose of Vladimir's suggestion.
> 
> Regarding vectorisation, `isNaN` is a simple comparison and can be easily auto-vectorised without help from intrinsics.
> 
> My speculation:
> 
> A native comparison such as `x != x` can be parsed directly by the compiler. As a result, the graph of the expression `if (x != x)` is simply
> 
> ```
> CmpF
>  |
> Bool
>  |
>  If
> ```
> 
> Your intrinsics, on the other hand, do not return the results on the flags, which leads to an extra comparison when using in conditions, `if(isNaN(x))` becomes
> 
> ```
> IsNaN    0
>     \   /
>     CmpI
>      |
>     Bool
>      |
>      If
> ```
> 
> In your benchmark, however, using this comparison to cmoving between 0 and 1 (`false` and `true`), the compiler recognised the pattern `x != 0 ? 0 : 1` with `x` having the type of `TypeInt::BOOL`. As a result, it reduces the graph into
> 
> ```
> IsNaN    1
>     \   /
>     XorI
> ```
> 
> Personally, I'm not into this implementation of intrinsics. FYI, gcc and clang both use sequences similar to `x != x` for `std::isnan`, `Math.abs(x) <= MAX_VALUE` for `std::isfinite` and `Math.abs(x) > MAX_VALUE` for `std::isinf`. The first one reduces to a single instruction `ucomiss x, x` so there is no reason to optimise further. The others are compiled down to 2 instructions each `vandpd t, x, [SIGN_ELIMINATE]; ucomiss t, [MAX_VALUE]`, so to optimise these further requires careful assessments.
> 
> If you feel comfortable I would suggest you build the graph for these intrinsics as
> 
> ```
>  X
>  |
> Bool   1    0
>    \   |   /
>      CMove
> ```
> 
> Then we can add ideal rules to `BoolNode` to recognise the patterns
> 
> ```
>  X
>  |
> Bool   1    0
>    \   |   /
>      CMove     0
>           \   /
>            CmpI
>             |
>            Bool
> ```
> 
> And reduce them to
> 
> ```
>  X
>  |
> Bool
> ```
> 
> With this, we can have the `Double::isInfinite` intrinsics compiled down to `vfpclass k, x; ktest k`, which is much more preferable. For non-AVX512DQ though I would prefer implementing them in Java similar to described above. Both abs and comparison nodes are not hard to be vectorised so it would not be a problem.
> 
> Thanks a lot.

As per the performance data, 
- for `isInfinite()`, the intrinsic is giving 60% speedup for `Float` (and 40% for `Double`) compared to the Java implementation.
- for` isFinite()`, the intrinsic is giving 20% speedup for `Float` (and 6% slowdown for `Double`) instead of the Java implementation (`return Math.abs(f) <= Float.MAX_VALUE`). 
- for `isNan()`, the flags fixup is skewing the results in favor of the intrinsic (5x for` Float `and 4x for `Double`) but you're saying that your patch will make this intrinsic approach unnecessary. 

I agree with your logic, but we also need to back it up with performance data. Could you please run my benchmark for `isNaN() `with your patch and provide the performance data? That way we can quantify the performance of your proposed optimization of `isNaN()`.

vamsi-parasa commented:
Hi Vladimir (@vnkozlov)

For 32bit, in the case of double, we see performance improvement using `vfpclasssd` instruction but **without** `vfpclassd`, we see **40% decrease** in performance for `isFinite()` compared to the original Java code. Below, is the code which implements the intrinsic using SSE.

Is it Ok to skip support for **non** `vfpclassd` for 32bit?

```
void C2_MacroAssembler::double_class_check_sse(int opcode, XMMRegister src, Register dst, Register temp, Register temp1) {
  int32_t POS_INF_HI = 0x7ff00000; // hi 32bits
  int32_t KILL_SIGN_MASK_HI = 0x7fffffff; // hi 32 bits

  pshuflw(src, src, 0x4e); //switch hi to lo
  movdl(temp, src);
  movl(temp1, KILL_SIGN_MASK_HI);
  andl(temp, temp1);
  movl(temp1, POS_INF_HI);
  cmpl(temp, temp1);
  switch (opcode) {
    case Op_IsFiniteD:
      setb(Assembler::below, dst);
      break;
    case Op_IsInfiniteD:
      setb(Assembler::equal, dst);
      break;
    case Op_IsNaND:
      setb(Assembler::above, dst);
      break;
    default:
      assert(false, "%s", NodeClassNames[opcode]);
  }
  andl(dst, 0xff);
}
```

merykitty commented:
@vamsi-parasa I modified your benchmark to emulate more use cases of these functions and run it on the baseline, #8525 with modified `isInfinite` (to use `Math.abs(v) > MAX_VALUE` instead) and this patch. The result is as follows, the source code and the assembly for the interesting parts will be shown later

                                                       Baseline         #8459           #8525
    Benchmark                             Mode  Cnt  Score   Error  Score   Error   Score   Error  Units
    FloatClassCheck.testIsFiniteBranch    avgt    5  2.522 ± 0.094  2.564 ± 0.187   2.512 ± 0.137  ns/op
    FloatClassCheck.testIsFiniteCMov      avgt    5  0.479 ± 0.014  0.786 ± 0.009   0.475 ± 0.005  ns/op
    FloatClassCheck.testIsFiniteStore     avgt    5  0.482 ± 0.010  0.603 ± 0.026   0.480 ± 0.006  ns/op
    FloatClassCheck.testIsInfiniteBranch  avgt    5  1.921 ± 0.043  1.778 ± 0.023   1.767 ± 0.039  ns/op
    FloatClassCheck.testIsInfiniteCMov    avgt    5  1.124 ± 0.045  0.787 ± 0.013   0.622 ± 0.019  ns/op
    FloatClassCheck.testIsInfiniteStore   avgt    5  1.195 ± 0.033  0.602 ± 0.015   0.625 ± 0.033  ns/op
    FloatClassCheck.testIsNaNBranch       avgt    5  1.896 ± 0.182  2.097 ± 0.216   1.725 ± 0.222  ns/op
    FloatClassCheck.testIsNaNCMov         avgt    5  2.956 ± 0.021  0.856 ± 0.003   0.390 ± 0.006  ns/op
    FloatClassCheck.testIsNaNStore        avgt    5  3.024 ± 0.071  0.741 ± 0.139   0.410 ± 0.008  ns/op

                                                        Baseline         #8459           #8525
    Benchmark                              Mode  Cnt  Score   Error  Score   Error  Score   Error  Units
    DoubleClassCheck.testIsFiniteBranch    avgt    5  2.566 ± 0.105  3.023 ± 0.117  2.603 ± 0.137  ns/op
    DoubleClassCheck.testIsFiniteCMov      avgt    5  0.481 ± 0.010  0.978 ± 0.011  0.485 ± 0.018  ns/op
    DoubleClassCheck.testIsFiniteStore     avgt    5  0.480 ± 0.012  0.943 ± 0.012  0.486 ± 0.011  ns/op
    DoubleClassCheck.testIsInfiniteBranch  avgt    5  1.907 ± 0.081  1.917 ± 0.065  1.808 ± 0.039  ns/op
    DoubleClassCheck.testIsInfiniteCMov    avgt    5  1.111 ± 0.028  0.982 ± 0.019  0.630 ± 0.017  ns/op
    DoubleClassCheck.testIsInfiniteStore   avgt    5  1.134 ± 0.011  0.944 ± 0.017  0.630 ± 0.009  ns/op
    DoubleClassCheck.testIsNaNBranch       avgt    5  1.926 ± 0.218  2.193 ± 0.045  1.767 ± 0.142  ns/op
    DoubleClassCheck.testIsNaNCMov         avgt    5  2.944 ± 0.020  1.047 ± 0.012  0.392 ± 0.009  ns/op
    DoubleClassCheck.testIsNaNStore        avgt    5  3.011 ± 0.065  0.946 ± 0.029  0.411 ± 0.004  ns/op

The source code for `FloatClassCheck`, that of `DoubleClassCheck` is similar

    RandomGenerator rng;
    static final int BUFFER_SIZE = 1024;
    float[] inputs;
    boolean[] storeOutputs;
    int[] cmovOutputs;
    int[] branchOutputs;

    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
    static int call() {
        return 1;
    }

    @Setup
    public void setup() {
        storeOutputs = new boolean[BUFFER_SIZE];
        cmovOutputs = new int[BUFFER_SIZE];
        branchOutputs = new int[BUFFER_SIZE];
        inputs = new float[BUFFER_SIZE];
        RandomGenerator rng = RandomGeneratorFactory.getDefault().create(0);
        float input;
        for (int i = 0; i < BUFFER_SIZE; i++) {
            if (i % 5 == 0) {
                input = (i%2 == 0) ? Float.NEGATIVE_INFINITY : Float.POSITIVE_INFINITY;
            }
            else if (i % 3 == 0) input = Float.NaN;
            else input = rng.nextFloat();
            inputs[i] = input;
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsFiniteStore() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            storeOutputs[i] = Float.isFinite(inputs[i]);
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsInfiniteStore() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            storeOutputs[i] = Float.isInfinite(inputs[i]);
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsNaNStore() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            storeOutputs[i] = Float.isNaN(inputs[i]);
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsFiniteCMov() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            cmovOutputs[i] = Float.isFinite(inputs[i]) ? 9 : 7;
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsInfiniteCMov() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            cmovOutputs[i] = Float.isInfinite(inputs[i]) ? 9 : 7;
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsNaNCMov() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            cmovOutputs[i] = Float.isNaN(inputs[i]) ? 9 : 7;
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsFiniteBranch() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            cmovOutputs[i] = Float.isFinite(inputs[i]) ? call() : 7;
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsInfiniteBranch() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            cmovOutputs[i] = Float.isInfinite(inputs[i]) ? call() : 7;
        }
    }

    @Benchmark
    @OperationsPerInvocation(BUFFER_SIZE)
    public void testIsNaNBranch() {
        for (int i = 0; i < BUFFER_SIZE; i++) {
            cmovOutputs[i] = Float.isNaN(inputs[i]) ? call() : 7;
        }
    }

The assembly of the interesting parts of the executions:

        FloatClassCheck::testIsFiniteBranch:
    Baseline, #8525:
    vandps  -0xd4791(%rip), %xmm0, %xmm1
    vucomiss        %xmm1, %xmm2
    jae     -0x77

    #8459
    vmovd   %xmm0, %r11d
    andl    $0x7fffffff, %r11d
    cmpl    $0x7f800000, %r11d
    setb    %r10b
    andl    $0xff, %r10d
    testl   %r10d, %r10d
    jne     -0x90

        FloatClassCheck::testIsFiniteCMov:
    Baseline, #8525:
    vandps  -0xcdcf6(%rip), %xmm6, %xmm7
    vucomiss        %xmm7, %xmm10
    movl    $0x9, %r10d
    cmovbl  %r14d, %r10d

    #8459:
    vmovd   %xmm4, %r9d
    andl    $0x7fffffff, %r9d
    cmpl    $0x7f800000, %r9d
    setb    %r8b
    andl    $0xff, %r8d
    testl   %r8d, %r8d
    movl    $0x7, %ebx
    cmovnel %r14d, %ebx
    
        FloatClassCheck::isFiniteStore:
    Baseline, #8525:
    vandps  -0xcfd74(%rip), %xmm3, %xmm3
    movl    $0x1, %r10d
    vucomiss        %xmm3, %xmm0
    cmovbl  %r9d, %r10d

    #8459:
    vmovd   %xmm6, %edi
    andl    $0x7fffffff, %edi
    cmpl    $0x7f800000, %edi
    setb    %dil
    andl    $0xff, %edi

        FloatClassCheck::isInfiniteBranch:
    Baseline:
    vucomiss        -0xc8(%rip), %xmm1
    jp      0x2
    je      0x20
    vucomiss        -0xd0(%rip), %xmm1
    nopl    (%rax,%rax)
    nop
    jp      -0x86
    jne     -0x8c

    #8459:
    vmovd   %xmm1, %r10d
    andl    $0x7fffffff, %r10d
    cmpl    $0x7f800000, %r10d
    sete    %r11b
    andl    $0xff, %r11d
    testl   %r11d, %r11d
    je      -0x87

    #8525:
    vandps  -0xce478(%rip), %xmm1, %xmm0
    nopl    (%rax,%rax)
    vucomiss        -0xc8(%rip), %xmm0
    jbe     -0x76

        FloatClassCheck::isInfiniteCMov:
    Baseline:
    vucomiss        -0x128(%rip), %xmm1
    jp      0x2
    je      0x16
    vucomiss        -0x130(%rip), %xmm1
    jp      0x2
    je      0xa

    #8459:
    vmovd   %xmm5, %eax
    andl    $0x7fffffff, %eax
    cmpl    $0x7f800000, %eax
    sete    %bpl
    andl    $0xff, %ebp
    testl   %ebp, %ebp
    movl    $0x7, %eax
    cmovnel %ebx, %eax

    #8525:
    vandps  -0xcefc3(%rip), %xmm0, %xmm0
    vucomiss        -0x12b(%rip), %xmm0
    movl    $0x9, %esi
    cmovbel %r8d, %esi

        FloatClassCheck::isInfiniteStore:
    Baseline:
    vucomiss        -0x128(%rip), %xmm0
    jp      0x2
    je      0x11
    vucomiss        -0x130(%rip), %xmm0
    jp      0x2
    je      0x5

    #8459:
    vmovd   %xmm2, %r8d
    andl    $0x7fffffff, %r8d
    cmpl    $0x7f800000, %r8d
    sete    %r8b
    andl    $0xff, %r8d

    #8525:
    vandps  -0xcf2b9(%rip), %xmm0, %xmm0
    vucomiss        -0x121(%rip), %xmm0
    movl    $0x1, %r11d
    cmovbel %esi, %r11d

        FloatClassCheck::isNaNBranch:
    Baseline:
    vucomiss        %xmm0, %xmm0
    jp      0x2
    je      -0x64

    #8459:
    vmovd   %xmm1, %r10d
    andl    $0x7fffffff, %r10d
    cmpl    $0x7f800000, %r10d
    seta    %r11b
    andl    $0xff, %r11d
    testl   %r11d, %r11d
    je      -0x87

    #8525:
    vucomiss        %xmm1, %xmm1
    jnp     -0x62

        FloatClassCheck::isNaNCMov:
    Baseline:
    vucomiss        %xmm5, %xmm5
    jnp     0xa
    pushfq
    andq    $-0xd5, (%rsp)
    popfq
    movl    $0x7, %r9d
    cmovnel %r8d, %r9d

    #8459:
    vmovd   %xmm4, %ebp
    andl    $0x7fffffff, %ebp
    cmpl    $0x7f800000, %ebp
    seta    %al
    andl    $0xff, %eax
    testl   %eax, %eax
    movl    $0x7, %ebp
    cmovnel %ebx, %ebp

    #8525:
    vucomiss        %xmm4, %xmm4
    movl    $0x7, %r9d
    cmovpl  %r8d, %r9d

        FloatClassCheck::isNaNStore:
    Baseline:
    vucomiss        %xmm3, %xmm3
    jnp     0xa
    pushfq
    andq    $-0xd5, (%rsp)
    popfq
    movl    $0x1, %ebx
    cmovel  %eax, %ebx

    #8459:
    vmovd   %xmm6, %edi
    andl    $0x7fffffff, %edi
    cmpl    $0x7f800000, %edi
    seta    %dil
    andl    $0xff, %edi

    #8525:
    movl    $0x1, %r9d
    vucomiss        %xmm0, %xmm0
    cmovnpl %r10d, %r9d

The assembly output for `DoubleClassCheck` is similar. Thanks.

vnkozlov commented:
> For 32bit, in the case of double, we see performance improvement using `vfpclasssd` instruction but **without** `vfpclassd`, we see **40% decrease** in performance for `isFinite()` compared to the original Java code. Below, is the code which implements the intrinsic using SSE.
> 
> Is it Ok to skip support for **non** `vfpclassd` for 32bit?

Yes, but add comment about that. Also for 32-bit you need to check SSE2 support which is required by `pshuflw`.

vamsi-parasa commented:
> > For 32bit, in the case of double, we see performance improvement using `vfpclasssd` instruction but **without** `vfpclassd`, we see **40% decrease** in performance for `isFinite()` compared to the original Java code. Below, is the code which implements the intrinsic using SSE.
> > Is it Ok to skip support for **non** `vfpclassd` for 32bit?
> 
> Yes, but add comment about that. Also for 32-bit you need to check SSE2 support which is required by `pshuflw`.

Thanks Vladimir! Will add a comment that the intrinsic doesn't give speedup without `vfpclasssd`. 
Yes, the predicate to check for `predicate(UseSSE>=2)` was added in the macro shown below.
```
instruct DoubleClassCheck_reg_reg_sse(rRegI dst, regD src, rRegI tmp, rRegI tmp1, rFlagsReg cr)
%{
  predicate(UseSSE>=2);
  match(Set dst (IsInfiniteD src));
  match(Set dst (IsNaND src));
  match(Set dst (IsFiniteD src));
  effect(TEMP tmp, TEMP tmp1, KILL cr);
  format %{ "double_class_check $dst, $src" %}
  ins_encode %{
    int opcode = this->ideal_Opcode();
    __ double_class_check_sse(opcode, $src$$XMMRegister, $dst$$Register, $tmp$$Register,
                $tmp1$$Register);
  %}
  ins_pipe(pipe_slow);
%}
```

vamsi-parasa commented:
> @vamsi-parasa I modified your benchmark to emulate more use cases of these functions and run it on the baseline, #8525 with modified `isInfinite` (to use `Math.abs(v) > MAX_VALUE` instead) and this patch. The result is as follows, the source code and the assembly for the interesting parts will be shown later
> 
> ```
>                                                    Baseline         #8459           #8525
> Benchmark                             Mode  Cnt  Score   Error  Score   Error   Score   Error  Units
> FloatClassCheck.testIsFiniteBranch    avgt    5  2.522 ± 0.094  2.564 ± 0.187   2.512 ± 0.137  ns/op
> FloatClassCheck.testIsFiniteCMov      avgt    5  0.479 ± 0.014  0.786 ± 0.009   0.475 ± 0.005  ns/op
> FloatClassCheck.testIsFiniteStore     avgt    5  0.482 ± 0.010  0.603 ± 0.026   0.480 ± 0.006  ns/op
> FloatClassCheck.testIsInfiniteBranch  avgt    5  1.921 ± 0.043  1.778 ± 0.023   1.767 ± 0.039  ns/op
> FloatClassCheck.testIsInfiniteCMov    avgt    5  1.124 ± 0.045  0.787 ± 0.013   0.622 ± 0.019  ns/op
> FloatClassCheck.testIsInfiniteStore   avgt    5  1.195 ± 0.033  0.602 ± 0.015   0.625 ± 0.033  ns/op
> FloatClassCheck.testIsNaNBranch       avgt    5  1.896 ± 0.182  2.097 ± 0.216   1.725 ± 0.222  ns/op
> FloatClassCheck.testIsNaNCMov         avgt    5  2.956 ± 0.021  0.856 ± 0.003   0.390 ± 0.006  ns/op
> FloatClassCheck.testIsNaNStore        avgt    5  3.024 ± 0.071  0.741 ± 0.139   0.410 ± 0.008  ns/op
> 
>                                                     Baseline         #8459           #8525
> Benchmark                              Mode  Cnt  Score   Error  Score   Error  Score   Error  Units
> DoubleClassCheck.testIsFiniteBranch    avgt    5  2.566 ± 0.105  3.023 ± 0.117  2.603 ± 0.137  ns/op
> DoubleClassCheck.testIsFiniteCMov      avgt    5  0.481 ± 0.010  0.978 ± 0.011  0.485 ± 0.018  ns/op
> DoubleClassCheck.testIsFiniteStore     avgt    5  0.480 ± 0.012  0.943 ± 0.012  0.486 ± 0.011  ns/op
> DoubleClassCheck.testIsInfiniteBranch  avgt    5  1.907 ± 0.081  1.917 ± 0.065  1.808 ± 0.039  ns/op
> DoubleClassCheck.testIsInfiniteCMov    avgt    5  1.111 ± 0.028  0.982 ± 0.019  0.630 ± 0.017  ns/op
> DoubleClassCheck.testIsInfiniteStore   avgt    5  1.134 ± 0.011  0.944 ± 0.017  0.630 ± 0.009  ns/op
> DoubleClassCheck.testIsNaNBranch       avgt    5  1.926 ± 0.218  2.193 ± 0.045  1.767 ± 0.142  ns/op
> DoubleClassCheck.testIsNaNCMov         avgt    5  2.944 ± 0.020  1.047 ± 0.012  0.392 ± 0.009  ns/op
> DoubleClassCheck.testIsNaNStore        avgt    5  3.011 ± 0.065  0.946 ± 0.029  0.411 ± 0.004  ns/op
> ```
> 
> The source code for `FloatClassCheck`, that of `DoubleClassCheck` is similar
> 
> ```
> RandomGenerator rng;
> static final int BUFFER_SIZE = 1024;
> float[] inputs;
> boolean[] storeOutputs;
> int[] cmovOutputs;
> int[] branchOutputs;
> 
> @CompilerControl(CompilerControl.Mode.DONT_INLINE)
> static int call() {
>     return 1;
> }
> 
> @Setup
> public void setup() {
>     storeOutputs = new boolean[BUFFER_SIZE];
>     cmovOutputs = new int[BUFFER_SIZE];
>     branchOutputs = new int[BUFFER_SIZE];
>     inputs = new float[BUFFER_SIZE];
>     RandomGenerator rng = RandomGeneratorFactory.getDefault().create(0);
>     float input;
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         if (i % 5 == 0) {
>             input = (i%2 == 0) ? Float.NEGATIVE_INFINITY : Float.POSITIVE_INFINITY;
>         }
>         else if (i % 3 == 0) input = Float.NaN;
>         else input = rng.nextFloat();
>         inputs[i] = input;
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsFiniteStore() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         storeOutputs[i] = Float.isFinite(inputs[i]);
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsInfiniteStore() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         storeOutputs[i] = Float.isInfinite(inputs[i]);
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsNaNStore() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         storeOutputs[i] = Float.isNaN(inputs[i]);
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsFiniteCMov() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         cmovOutputs[i] = Float.isFinite(inputs[i]) ? 9 : 7;
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsInfiniteCMov() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         cmovOutputs[i] = Float.isInfinite(inputs[i]) ? 9 : 7;
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsNaNCMov() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         cmovOutputs[i] = Float.isNaN(inputs[i]) ? 9 : 7;
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsFiniteBranch() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         cmovOutputs[i] = Float.isFinite(inputs[i]) ? call() : 7;
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsInfiniteBranch() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         cmovOutputs[i] = Float.isInfinite(inputs[i]) ? call() : 7;
>     }
> }
> 
> @Benchmark
> @OperationsPerInvocation(BUFFER_SIZE)
> public void testIsNaNBranch() {
>     for (int i = 0; i < BUFFER_SIZE; i++) {
>         cmovOutputs[i] = Float.isNaN(inputs[i]) ? call() : 7;
>     }
> }
> ```
> 
> The assembly of the interesting parts of the executions:
> 
> ```
>     FloatClassCheck::testIsFiniteBranch:
> Baseline, #8525:
> vandps  -0xd4791(%rip), %xmm0, %xmm1
> vucomiss        %xmm1, %xmm2
> jae     -0x77
> 
> #8459
> vmovd   %xmm0, %r11d
> andl    $0x7fffffff, %r11d
> cmpl    $0x7f800000, %r11d
> setb    %r10b
> andl    $0xff, %r10d
> testl   %r10d, %r10d
> jne     -0x90
> 
>     FloatClassCheck::testIsFiniteCMov:
> Baseline, #8525:
> vandps  -0xcdcf6(%rip), %xmm6, %xmm7
> vucomiss        %xmm7, %xmm10
> movl    $0x9, %r10d
> cmovbl  %r14d, %r10d
> 
> #8459:
> vmovd   %xmm4, %r9d
> andl    $0x7fffffff, %r9d
> cmpl    $0x7f800000, %r9d
> setb    %r8b
> andl    $0xff, %r8d
> testl   %r8d, %r8d
> movl    $0x7, %ebx
> cmovnel %r14d, %ebx
> 
>     FloatClassCheck::isFiniteStore:
> Baseline, #8525:
> vandps  -0xcfd74(%rip), %xmm3, %xmm3
> movl    $0x1, %r10d
> vucomiss        %xmm3, %xmm0
> cmovbl  %r9d, %r10d
> 
> #8459:
> vmovd   %xmm6, %edi
> andl    $0x7fffffff, %edi
> cmpl    $0x7f800000, %edi
> setb    %dil
> andl    $0xff, %edi
> 
>     FloatClassCheck::isInfiniteBranch:
> Baseline:
> vucomiss        -0xc8(%rip), %xmm1
> jp      0x2
> je      0x20
> vucomiss        -0xd0(%rip), %xmm1
> nopl    (%rax,%rax)
> nop
> jp      -0x86
> jne     -0x8c
> 
> #8459:
> vmovd   %xmm1, %r10d
> andl    $0x7fffffff, %r10d
> cmpl    $0x7f800000, %r10d
> sete    %r11b
> andl    $0xff, %r11d
> testl   %r11d, %r11d
> je      -0x87
> 
> #8525:
> vandps  -0xce478(%rip), %xmm1, %xmm0
> nopl    (%rax,%rax)
> vucomiss        -0xc8(%rip), %xmm0
> jbe     -0x76
> 
>     FloatClassCheck::isInfiniteCMov:
> Baseline:
> vucomiss        -0x128(%rip), %xmm1
> jp      0x2
> je      0x16
> vucomiss        -0x130(%rip), %xmm1
> jp      0x2
> je      0xa
> 
> #8459:
> vmovd   %xmm5, %eax
> andl    $0x7fffffff, %eax
> cmpl    $0x7f800000, %eax
> sete    %bpl
> andl    $0xff, %ebp
> testl   %ebp, %ebp
> movl    $0x7, %eax
> cmovnel %ebx, %eax
> 
> #8525:
> vandps  -0xcefc3(%rip), %xmm0, %xmm0
> vucomiss        -0x12b(%rip), %xmm0
> movl    $0x9, %esi
> cmovbel %r8d, %esi
> 
>     FloatClassCheck::isInfiniteStore:
> Baseline:
> vucomiss        -0x128(%rip), %xmm0
> jp      0x2
> je      0x11
> vucomiss        -0x130(%rip), %xmm0
> jp      0x2
> je      0x5
> 
> #8459:
> vmovd   %xmm2, %r8d
> andl    $0x7fffffff, %r8d
> cmpl    $0x7f800000, %r8d
> sete    %r8b
> andl    $0xff, %r8d
> 
> #8525:
> vandps  -0xcf2b9(%rip), %xmm0, %xmm0
> vucomiss        -0x121(%rip), %xmm0
> movl    $0x1, %r11d
> cmovbel %esi, %r11d
> 
>     FloatClassCheck::isNaNBranch:
> Baseline:
> vucomiss        %xmm0, %xmm0
> jp      0x2
> je      -0x64
> 
> #8459:
> vmovd   %xmm1, %r10d
> andl    $0x7fffffff, %r10d
> cmpl    $0x7f800000, %r10d
> seta    %r11b
> andl    $0xff, %r11d
> testl   %r11d, %r11d
> je      -0x87
> 
> #8525:
> vucomiss        %xmm1, %xmm1
> jnp     -0x62
> 
>     FloatClassCheck::isNaNCMov:
> Baseline:
> vucomiss        %xmm5, %xmm5
> jnp     0xa
> pushfq
> andq    $-0xd5, (%rsp)
> popfq
> movl    $0x7, %r9d
> cmovnel %r8d, %r9d
> 
> #8459:
> vmovd   %xmm4, %ebp
> andl    $0x7fffffff, %ebp
> cmpl    $0x7f800000, %ebp
> seta    %al
> andl    $0xff, %eax
> testl   %eax, %eax
> movl    $0x7, %ebp
> cmovnel %ebx, %ebp
> 
> #8525:
> vucomiss        %xmm4, %xmm4
> movl    $0x7, %r9d
> cmovpl  %r8d, %r9d
> 
>     FloatClassCheck::isNaNStore:
> Baseline:
> vucomiss        %xmm3, %xmm3
> jnp     0xa
> pushfq
> andq    $-0xd5, (%rsp)
> popfq
> movl    $0x1, %ebx
> cmovel  %eax, %ebx
> 
> #8459:
> vmovd   %xmm6, %edi
> andl    $0x7fffffff, %edi
> cmpl    $0x7f800000, %edi
> seta    %dil
> andl    $0xff, %edi
> 
> #8525:
> movl    $0x1, %r9d
> vucomiss        %xmm0, %xmm0
> cmovnpl %r10d, %r9d
> ```
> 
> The assembly output for `DoubleClassCheck` is similar. Thanks.

Thanks for sharing the performance data. Your patch is showing` ~2.5x` improvement over the intrinsic for the case of `{Float/Double}.testIsNaNCMov`.

vamsi-parasa commented:
Hi Vladimir (@vnkozlov),

Could you pls review this updated PR? In this updated patch, we **removed** the intrinsics using **non**`-vpfclasss/d` instructions.

- Got the new performance data for `vfpclasss/d`  intrinsics, after rebasing with the latest changes (which include #8525 submitted by @merykitty). 
- Using `vfpclasss/d` instruction gives upto `70%` speedup over the existing baseline.   

Please see the updated data shown below (also updated the RFE main text as well)

```
Benchmark (ns/op)	        Baseline Intrinsic(vfpclasss/d)	Speedup(%)
FloatClassCheck.testIsFinite	0.562	          0.406	         28%
FloatClassCheck.testIsInfinite	0.815	          0.383	         53%
FloatClassCheck.testIsNaN	0.63	          0.382	         39%
DoubleClassCheck.testIsFinite	0.565	          0.409	         28%
DoubleClassCheck.testIsInfinite	0.812	          0.375	         54%
DoubleClassCheck.testIsNaN	0.631	          0.38	         40%
FPComparison.isFiniteDouble	332.638	          272.577	 18%
FPComparison.isFiniteFloat	413.217	          331.825	 20%
FPComparison.isInfiniteDouble	874.897	          240.632	 72%
FPComparison.isInfiniteFloat	872.279	          321.269	 63%
FPComparison.isNanDouble	286.566	          240.36	 16%
FPComparison.isNanFloat	        346.123	          316.923	  8%
```


Thanks,
Vamsi

vnkozlov commented on src/hotspot/share/runtime/vmStructs.cpp:
```diff
@@ -1844,6 +1844,7 @@
   declare_c2_type(CopySignFNode, Node)                                    \
   declare_c2_type(SignumDNode, Node)                                      \
   declare_c2_type(SignumFNode, Node)                                      \
+  declare_c2_type(IsInfiniteFNode, Node)                                  \
```
Where are other new nodes?

vamsi-parasa replied:
I did not know how to combine the 3 nodes into one. Could you pls give few pointers to refer to for implementing them?

vamsi-parasa replied:
Fixed it and pushed a new update. Sorry, that got missed!

vnkozlov replied:
Good.

vnkozlov reviewed (commented):
Looks good.

Please, adopt @merykitty suggested changes for micro-benchmarks to include `store, cmove, branch` cases. Show performance results for all of them.

Also adapt your regression tests for all 3 cases too.

vnkozlov commented:
I assume `Baseline` data includes #8525 changes. Right?

vamsi-parasa commented:
> I assume `Baseline` data includes #8525 changes. Right?

Yes, the basline data includes #8525 changes.

vamsi-parasa commented:
> Looks good.
> 
> Please, adopt @merykitty suggested changes for micro-benchmarks to include `store, cmove, branch` cases. Show performance results for all of them.
> 
> Also adapt your regression tests for all 3 cases too.

Sure, I will update the microbenchmarks to include `store, cmove, branch` cases and post the performance data.

vamsi-parasa commented:
Hello Validmir (@vnkozlov),

After reviewing the JMH performance data, it was observed that only the` isInfinite()` method sees performance benefit using intrinsics (`vfpclassss/d` instruction). Thus,  this PR is updated to support intrinsics for **only** the `isInfinite()` method. 

Please see the performance data below for all the 3 methods. Please let me know if anything else is needed.

```
Benchmark (ns/op)	              Baseline 	Intrinsic  Speedup (%)
                                               (vfpclassss)
-------------------------------------------------------------------
FloatClassCheck.testIsFiniteBranch	1.558	1.535	     1%
FloatClassCheck.testIsFiniteCMov	0.335	0.428	    -28%
FloatClassCheck.testIsFiniteStore	0.417	0.253	     39%
---------------------------------------------------------------------
FloatClassCheck.testIsInfiniteBranch	1.294	1.046	     19%
FloatClassCheck.testIsInfiniteCMov	0.823	0.351	     57%
FloatClassCheck.testIsInfiniteStore	0.748	0.234	     69%
----------------------------------------------------------------------
FloatClassCheck.testIsNaNBranch	          1	1.147	    -15%
FloatClassCheck.testIsNaNCMov	        0.297	0.352       -19%
FloatClassCheck.testIsNaNStore	        0.362	0.234	     35%

*********************************************************************

Benchmark (ns/op)	              Baseline 	Intrinsic  Speedup (%)
                                               (vfpclasssd)
------------------------------------------------------------------------
DoubleClassCheck.testIsFiniteBranch	1.555	1.522	     2%
DoubleClassCheck.testIsFiniteCMov	0.325	0.444	    -37%
DoubleClassCheck.testIsFiniteStore	0.322	0.266	    17%
-----------------------------------------------------------------------
DoubleClassCheck.testIsInfiniteBranch	1.261	1.094	    13%
DoubleClassCheck.testIsInfiniteCMov	0.821	0.373	    55%
DoubleClassCheck.testIsInfiniteStore	0.858	0.235	    73%
-----------------------------------------------------------------------
DoubleClassCheck.testIsNaNBranch	1	1.127	   -13%
DoubleClassCheck.testIsNaNCMov	        0.278	0.373	   -34%
DoubleClassCheck.testIsNaNStore	        0.283	0.235	    17%

```

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -10136,3 +10142,31 @@ instruct castVVLeg(legVec dst)
+
+instruct FloatClassCheck_reg_reg_vfpclass(rRegI dst, regF src, kReg ktmp, rFlagsReg cr)
+%{
+  predicate(VM_Version::supports_avx512dq());
```
Predicate is no longer needed you have done the check in match_rule_supported routine.

vamsi-parasa replied:
That's true, the check is already done in the match rule. Will update the code.

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -10136,3 +10142,31 @@ instruct castVVLeg(legVec dst)
+
+instruct DoubleClassCheck_reg_reg_vfpclass(rRegI dst, regD src, kReg ktmp, rFlagsReg cr)
+%{
+  predicate(VM_Version::supports_avx512dq());
```
Same as above

jatin-bhateja commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -5316,3 +5316,31 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
+}
+
+void C2_MacroAssembler::double_class_check_vfp(int opcode, Register dst, XMMRegister src, KRegister tmp) {
+  uint8_t imm8;
```
May be ok to move it back to instruction encoding block , only two instructions.

vamsi-parasa replied:
That's true. The two instructions can be put in the instruction encoding block. Will do that.

vamsi-parasa replied:
Also, in future if we want to add the support for isFinite() and IsNaN(), wouldn't it be better to have seperate macros?

vamsi-parasa replied:
Updated the code with the suggested changes. Thanks Jatin and Vladimir!

vamsi-parasa commented:
> @vamsi-parasa can you show difference in generated code for `DoubleClassCheck.testIsFiniteCMov ` for example? Or may be for all of them (for DoubleClassCheck). I am fine with intrinsifying only `isInfinite()` but would like to see code for the record.

Sure Vladimir, will post the generated code of all the case of DoubleClassCheck.

vamsi-parasa commented:
> @vamsi-parasa can you show difference in generated code for `DoubleClassCheck.testIsFiniteCMov ` for example? Or may be for all of them (for DoubleClassCheck). I am fine with intrinsifying only `isInfinite()` but would like to see code for the record.

Hi Vladimir (@vnkozlov), please see the generated code below.

`isInfinite()` intrinsic is faster because the baseline produces` 2 vucomisd` instructions to compare with positive and negative infinity whereas the intrinsic generates only one `vfpclasssd` instruction as seen below:

**DoubleClassCheck.testIsInfiniteBranch (baseline)**
                                                                 
```
vucomisd -0xad(%rip),%xmm0         
jp     0x00007fe1aced1f31
je     0x00007fe1aced1f3d          
vucomisd -0xb1(%rip),%xmm0      
jp     0x00007fe1aced1ee0
jne    0x00007fe1aced1ee0
```
**DoubleClassCheck.testIsInfiniteBranch (intrinsic/vfpclasssd)**
```
vfpclasssd $0x18,%xmm0,%k7
kmovb  %k7,%edi
test   %edi,%edi
je     0x00007f1270c167e0 
```

-------------------------
However, in the case of` isFinite() `and `isNan()`, only `1 vucomisd` is generated and is faster for `Cmov` and` branch/call` tests as seen below

**DoubleClassCheck.testIsFiniteCmov (baseline)**
 
```
vandpd -0x141493(%rip),%xmm0,%xmm0        
vucomisd %xmm0,%xmm1
mov    $0x9,%ecx
cmovb  %r8d,%ecx

```
**DoubleClassCheck.testIsFiniteCmov (intrinsci/vfpclasssd)**

```
vfpclasssd $0x99,%xmm2,%k7
kmovb  %k7,%r11d
xor    $0x1,%r11d
test   %r11d,%r11d
mov    $0x7,%r11d
cmovne %r8d,%r11d
```
-------------------------------------------------------------------------

**DoubleClassCheck.testIsNaNCmov(baseline)**
 
```
vucomisd %xmm0,%xmm0
jp     0x00007f0a4d1425fb           
movl   $0x7,0x10(%rcx,%rax,4) 
```

**DoubleClassCheck.testIsNaNCmov (intrinsic/vfpclasssd)**
 
```
vfpclasssd $0x81,%xmm0,%k7
kmovb  %k7,%r8d                    
test   %r8d,%r8d
jne  0x00007f7fa0c50f43    
movl   $0x7,0x10(%r10,%rdx,4) 
```  

Please let me know if further data is needed. 

Thanks,
Vamsi

vnkozlov reviewed (approved):
Looks good. Let me test it before you push.

vamsi-parasa commented:
> Looks good. Let me test it before you push.

Thanks Vladimir! Will wait until the tests pass...

vnkozlov commented:
compiler/c2/irTests/TestScheduleSmallMethod.java failed in Tier1 (case #1 is run with "-XX:-OptoScheduling"):

```
compiler.lib.ir_framework.shared.TestRunException: The following scenarios have failed: #1. Please check stderr for more information.
	at compiler.lib.ir_framework.TestFramework.reportScenarioFailures(TestFramework.java:617)
	at compiler.lib.ir_framework.TestFramework.startWithScenarios(TestFramework.java:578)
	at compiler.lib.ir_framework.TestFramework.start(TestFramework.java:335)
	at compiler.c2.irTests.TestScheduleSmallMethod.main(TestScheduleSmallMethod.java:44)
```

The test use Doube arithmetic.

vamsi-parasa commented:
> On other hand it failed on Windows version we have issue with recently. Several IR framework tests failed with "Did not find IR encoding": [8286979](https://bugs.openjdk.java.net/browse/JDK-8286979)
> 
> Also it run on AMD CPU which does not have EVEX and AVX512. Your code should not be executed.
> 
> False alarm it seems. I will run other tiers.

Will update the tests not to run on AMD CPUs which don't EVEX and AVX512. Thanks!

vnkozlov commented:
> > On other hand it failed on Windows version we have issue with recently. Several IR framework tests failed with "Did not find IR encoding": [8286979](https://bugs.openjdk.java.net/browse/JDK-8286979)
> > Also it run on AMD CPU which does not have EVEX and AVX512. Your code should not be executed.
> > False alarm it seems. I will run other tiers.
> 
> Will update the tests not to run on AMD CPUs which don't EVEX and AVX512. Thanks!

I don't think you need update any tests. Your new tests already have: `@requires vm.cpu.features ~= ".*avx512dq.*"`

vamsi-parasa commented:
> > > On other hand it failed on Windows version we have issue with recently. Several IR framework tests failed with "Did not find IR encoding": [8286979](https://bugs.openjdk.java.net/browse/JDK-8286979)
> > > Also it run on AMD CPU which does not have EVEX and AVX512. Your code should not be executed.
> > > False alarm it seems. I will run other tiers.
> > 
> > 
> > Will update the tests not to run on AMD CPUs which don't EVEX and AVX512. Thanks!
> 
> I don't think you need update any tests. Your new tests already have: `@requires vm.cpu.features ~= ".*avx512dq.*"`

Got it!

vnkozlov reviewed (approved):
Testing results are good.

You need second review approval.

vamsi-parasa commented:
> Testing results are good.
> 
> You need second review approval.

Thanks Vladimir! Will ask for second review/approval.
For the github pre-submit tests, on Windows x86, HighlightUITest.java is failing.

vnkozlov commented:
> > Testing results are good.
> > You need second review approval.
> 
> Thanks Vladimir! Will ask for second review/approval. For the github pre-submit tests, on Windows x86 (Tier 1), HighlightUITest.java is failing.

known issue: [JDK-8284144](https://bugs.openjdk.java.net/browse/JDK-8284144)

vamsi-parasa commented:
> compiler/c2/irTests/TestScheduleSmallMethod.java failed in Tier1 (case #1 is run with "-XX:-OptoScheduling"):
> 
> ```
> compiler.lib.ir_framework.shared.TestRunException: The following scenarios have failed: #1. Please check stderr for more information.
> 	at compiler.lib.ir_framework.TestFramework.reportScenarioFailures(TestFramework.java:617)
> 	at compiler.lib.ir_framework.TestFramework.startWithScenarios(TestFramework.java:578)
> 	at compiler.lib.ir_framework.TestFramework.start(TestFramework.java:335)
> 	at compiler.c2.irTests.TestScheduleSmallMethod.main(TestScheduleSmallMethod.java:44)
> ```
> 
> The test use Doube arithmetic.

Hi Vladimir, wanted to check with you if this test is still a problem and needs to be fixed?

vamsi-parasa commented:
/integrate

jatin-bhateja commented:
/sponsor
